### PR TITLE
Removed RuntimeQueue

### PR DESCRIPTION
**Pull request by dVakulen**
Considered ```RuntimeQueue``` as dead code: constant ```USE_BLOCKING_COLLECTION ``` was not changed for the last year.

### CODE REVIEW

**Comment by gabikliot** [comment]
I would suggest against removing it. It is a wrapper for various different implementations of the queue. And we did experiment with that in the past. As of now, `BlockingCollection` seems to be the better one, but I am not convinced another even better one could not be found. And anyway, this looks like a good abstraction to have anyway. There might be more behavior we might want to add to it going forward. This is sort of a basic building block in the runtime, so its good to be abstracted I think.

**Comment by dVakulen** [comment]
@gabikliot Yes, I thought about it too, and already experimented with ```ConcurrentBag``` underlying collection and it's using won't result in any performance increase (performance comparison of different concurrent collections can be seen at http://blogs.msdn.com/b/pfxteam/archive/2010/04/26/performance-of-concurrent-collections-in-net-4.aspx). Secondly, if even there's still experiments to be done with different collections - I'm not sure, that 1 concrete collection will benefit every place where ```RuntimeQueue``` is being used. For example on local tests I've seen the following results:
image
where DoWork() had workload of about 500 FLOPs. And ```SpinWait.SpinOnce``` method is taking almost same amount of CPU. Besides that lock contention seems pretty high.
Taking into account all of the above I think that experiments are definitely needed, but probably more cardinal ones than just swapping collection type.

**Comment by gabikliot** [comment]
>I think that experiments are definitely needed, but probably more cardinal ones than just changing collection type.

Totally agree!
And especially given what you wrote that one solution might not fit it all, we better have an abstraction for it, and the `RuntimeQueue` can be created with different behavior inside, so can be specialized for different cases.

The question we are after here: do we need the abstraction or not? I think we better keep it. Do you see any downsides?

**Comment by dVakulen** [comment]
For now it's suits our needs, so it should be kept. And when advanced usage of collection will require us to add method which is not supported by another collections it will be easy to change ```RuntimeQueue``` to concrete type. So now question is whether handmade through ```lock```'ing  concurrent collections will be needed or not.

**Comment by gabikliot** [comment]
NO harm in keeping it, right?
BTW, from the link you shared, NaiveBlockingQueue<T> actually outperforms the BlockingCollection  at low load.

Also, re spin locks: what we have seen with all those concurrent collections in .NET that use spin locks, is that spin locks are "elastic" - under light load they will consume a lot of CPU, so it will appear as if they are wasteful. But once you increase the load, the queues protected by spin locks are full, so they don't spin actually. So as load increases we spend LESS time in spin locks.
The outcome is pretty fascinating and unexpected:
lets says you push X load (X requests per second) and CPU is at 70%. So you think you can maybe only push 1.5X load until CPU is maxed. But you push 2X or maybe even 2.5X load and CPU goes only to 90%. **CPU does not increase linearly with the load, due to non-linearity of spin locks.**
 
The bottom line: its pretty hard to predict what will work better. The only way is to abstract the implementation, try different ways and measure. And that is what this `RuntimeQueue` was about.

**Comment by ReubenBond** [comment]
I think we should remove unused code and keep the codebase as clean, lean, and approachable as possible.

Git records the history, so nothing is lost. We can put good but unrequired code like this in a separate repository.

**Comment by sergeybykov** [comment]
I agree with @ReubenBond - let's keep the codebase clean. Nothing is lost in the source control anyway.

**Comment by jdom** [comment]
@gabikliot I was convinced by your argument that spinning a lot under light load is OK, since you don't have load anyway. In most cases that's fine, assuming that the Silo is the ONLY thing in your VM or physical machine. But I've seen more and more people willing to deploy multiple silos in a single machine (microservices?) or even the frontend and the Silo... I understand this is discouraged, but they still do it, so we should try to avoid a little but of the busy work under light load if we can avoid it.
Aside from deploying multiple silos or a web apps, there's also other valid processes that could be running alongside the Silo (for example logging agents that work out of process, mentioning this example, as in some 343 services the CPU overhead that the log agent has is not at all negligible).

Maybe we should revisit experimenting with collections or at least approaches under light load at some point.

**Comment by gabikliot** [comment]
Very good point @jdom!
I agree.

Therefore, the steps I would suggest are:
1) Put back the `RuntimeQueue` abstraction that was removed in this PR (despite my opinion), exactly for the purpose you explained above: so we can experiment and eventually allow to configure different implementations of this abstraction.
2) Put back the lock based implementation, as one of the options.
3) Add config variable, to allow to configure different runtime queues (that's what we were missing before).

That alone already takes you quite a lot towards the goal you outlined above.
I still recommend the default to be Blocking Collection, like now, and it will work out of the box for the majority of users. But more advanced users will be able to optimize their scenario via more advanced configuration.

4) The next step can be to explore other implementations, or maybe some smart combination of the implementations when you dynamically switch from one to another. I would be interested to follow up with you on any thoughts and brainstorming that you might have. You have to know that we did spend quite a considerable time trying different options (I had an intern one summer with whom we tried different options), and we were not able to find something that will beat the current implementation on our target benchmark (Halo Heartbeat stress test). But I can easily admit that we probably did not try hard, and were also not looking into targeting multiple usage scenarios, like you described above. Therefore, I do encourage you/others to try again, harder.

**Comment by ReubenBond** [comment]
> 3) Add config variable, to allow to configure different runtime queues (that's what we were missing before).

I'd prefer we use dependency injection over a config variable.
